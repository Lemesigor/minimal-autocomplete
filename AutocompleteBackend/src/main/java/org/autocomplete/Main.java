package org.autocomplete;

import org.autocomplete.application.TermService;
import org.autocomplete.infrastructure.InMemoryTermsRepository;
import org.autocomplete.infrastructure.controller.TermsController;
import org.autocomplete.infrastructure.jdbc.ConnectionFactory;
import org.autocomplete.infrastructure.jdbc.TermJdbcRepository;
import org.autocomplete.infrastructure.rest.HttpClient;
import org.autocomplete.infrastructure.rest.SparkClient;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import spark.Route;

import java.sql.Connection;
import java.sql.SQLException;


public class Main {
    public static void main(String[] args) {
        Logger logger = LoggerFactory.getLogger(Main.class);
        logger.info("Initialing api");

        try {
            Connection connection = ConnectionFactory.createConnection(
                    "localhost",
                    "root",
                    "root",
                    "autocomplete",
                    "3306"
            );

            HttpClient<Route> client = new SparkClient();
            TermJdbcRepository repository = new TermJdbcRepository(connection);
            TermService service = new TermService(repository);
            TermsController controller = new TermsController(client, service);

            controller.listen();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }
}