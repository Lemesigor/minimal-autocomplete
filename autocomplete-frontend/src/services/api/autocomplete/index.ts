const BASE_ENDPOINT = "https://localhost:4567/autocomplete?";

const DEFAULT_LIMIT = 10;

export type Term = {
    id: number;
    term: string;
}

export type TermResponse = {
    terms: Term[];
    total: number;
}

export const getAutocompleteData = async (query: string = "", signal: AbortSignal): Promise<Term[]> => {
    try {
        const autocompleteEndpoint = `${BASE_ENDPOINT}query=${encodeURIComponent(query)}&limit=${DEFAULT_LIMIT}`;
        const response = await fetch(autocompleteEndpoint, { signal });
        if (!response.ok) {
            console.error(`HTTP error! status: ${response.status}`);
            return [];
        }
        const data: TermResponse = await response.json();
        return data?.terms || []; // Return terms instead of results
    } catch (error) {
        // @ts-ignore
        if (error.name === "AbortError") {
            // @ts-ignore
            console.log("Request aborted:", error.message);
        } else {
            // @ts-ignore
            console.error("Request error:", error.message);
        }
        return [];
    }
};
