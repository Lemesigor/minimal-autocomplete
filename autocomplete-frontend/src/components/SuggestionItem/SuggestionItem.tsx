/** @jsxImportSource @emotion/react */
import {css} from '@emotion/react';
import {Term} from "../../services/api/autocomplete";

type SuggestionItemProps = {
    item: Term;
    handleClick: (item: Term) => void;
    isRecentSearch: boolean;
};

const itemStyle = css`
    display: flex;
    align-items: center;
    background-color: rgba(255, 255, 255, 0.5);
    transition: all 0.2s ease;
    padding: 0.25rem 0;

    &:hover {
        background-color: rgba(226, 232, 240, 1);
    }
`;

const iconStyle = css`
    margin-right: 0.5rem;
    height: 15px;
    height: 15px;
`;


export const SuggestionItem = ({item, handleClick, isRecentSearch}: SuggestionItemProps) => {
    const iconPath = isRecentSearch ? "recent_search.png" : "search.png";
    return (
        <li
            tabIndex={-1}
            key={item.id}
            css={itemStyle}
            onMouseDown={(e) => e.preventDefault()}
            onClick={() => handleClick(item)}
        >
            <img src={iconPath} css={iconStyle} alt={"search"}/>
            <span className="cursor-default">{item.value}</span>
        </li>
    )
};

