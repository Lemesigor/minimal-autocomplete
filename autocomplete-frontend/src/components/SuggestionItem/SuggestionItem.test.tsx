/** @jsxImportSource @emotion/react */
import {render, fireEvent} from "@testing-library/react";
import {SuggestionItem} from "./SuggestionItem";
import {Term} from "../../services/api/autocomplete";

describe("SuggestionItem", () => {
    const mockTerm: Term = {id: 1, value: "test"};
    const mockHandleClick = jest.fn();

    it("should match the snapshot with search", () => {
        const {container} = render(<SuggestionItem item={mockTerm} handleClick={mockHandleClick}
                                                   isRecentSearch={true}/>);
        expect(container).toMatchSnapshot();
    });

    it("should match the snapshot without search", () => {
        const {container} = render(<SuggestionItem item={mockTerm} handleClick={mockHandleClick}
                                                   isRecentSearch={false}/>);
        expect(container).toMatchSnapshot();
    });

    it("should call handleClick when the item is clicked", () => {
        const {getByText} = render(<SuggestionItem item={mockTerm} handleClick={mockHandleClick}
                                                   isRecentSearch={false}/>);
        const itemElement = getByText("test");
        fireEvent.click(itemElement);
        expect(mockHandleClick).toHaveBeenCalledTimes(1);
        expect(mockHandleClick).toHaveBeenCalledWith(mockTerm);
    });
});