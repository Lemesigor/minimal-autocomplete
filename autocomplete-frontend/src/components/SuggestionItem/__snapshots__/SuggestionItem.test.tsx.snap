// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SuggestionItem should match the snapshot with search 1`] = `
<div>
  <li
    class="css-r1srfx"
    tabindex="-1"
  >
    <img
      alt="search"
      class="css-k0ltuw"
      src="recent_search.png"
    />
    <span
      class="cursor-default"
    >
      test
    </span>
  </li>
</div>
`;

exports[`SuggestionItem should match the snapshot without search 1`] = `
<div>
  <li
    class="css-r1srfx"
    tabindex="-1"
  >
    <img
      alt="search"
      class="css-k0ltuw"
      src="search.png"
    />
    <span
      class="cursor-default"
    >
      test
    </span>
  </li>
</div>
`;
